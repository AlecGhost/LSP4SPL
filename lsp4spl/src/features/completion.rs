use crate::document::DocumentRequest;
use color_eyre::eyre::Result;
use lsp_types::{CompletionItem, CompletionItemKind, CompletionParams, Documentation, MarkupKind};
use spl_frontend::{
    ast::{GlobalDeclaration, ProcedureDeclaration, Statement, TypeDeclaration},
    lexer::token::{Token, TokenList, TokenType},
    table::{
        GlobalEntry, GlobalTable, LocalEntry, LocalTable, LookupTable, SymbolTable, TableEntry,
    },
    ToRange,
};
use tokio::sync::mpsc::Sender;

pub async fn propose(
    doctx: Sender<DocumentRequest>,
    params: CompletionParams,
) -> Result<Option<Vec<CompletionItem>>> {
    let doc_params = params.text_document_position;
    if let Some(cursor) = super::doc_cursor(doc_params, doctx).await? {
        let table = &cursor.doc_info.table;
        let program = &cursor.doc_info.ast;
        let position = correct_index(cursor.index);

        if let Some(gd) = program
            .global_declarations
            .iter()
            .find(|gd| gd.to_range().contains(&position))
        {
            use GlobalDeclaration::*;
            match gd {
                Type(td) => Ok(complete_type(td, position, table)),
                Procedure(pd) => Ok(complete_procedure(pd, position, table)),
                Error(_) => Ok(None),
            }
        } else {
            // fixes, that a cursor at the end of an unfinished type declaration at the end of the
            // file is still recognized as part of the declaration
            if let Some(GlobalDeclaration::Type(td)) = program.global_declarations.last() {
                if let Some(last_token) = td.info.tokens.last() {
                    if !matches!(last_token.token_type, TokenType::Semic) {
                        return Ok(complete_type(td, position, table));
                    }
                }
            }
            // The cursor is not inside the scope of any global declaration,
            // which means, that a new one can be startet with the `proc` or `type` keywords.
            // Furthermore, some snippets are provided for those definitions
            let mut completions = vec![
                snippets::proc(),
                snippets::r#type(),
                items::proc(),
                items::r#type(),
            ];
            if let Some(GlobalEntry::Procedure(_)) = cursor.doc_info.table.lookup("main") {
                // main already exists
            } else {
                completions.push(snippets::main());
            }
            Ok(Some(completions))
        }
    } else {
        Ok(None)
    }
}

fn complete_type(
    td: &TypeDeclaration,
    position: usize,
    table: &GlobalTable,
) -> Option<Vec<CompletionItem>> {
    td.info
        .tokens
        .token_before(position)
        .and_then(|last_token| {
            use TokenType::*;
            match last_token.token_type {
                Eq => Some(vec![snippets::array(), items::array(), items::int()]),
                RBracket => Some(vec![items::of()]),
                Of => {
                    let completions =
                        vec![vec![snippets::array(), items::array()], search_types(table)].concat();
                    Some(completions)
                }
                _ => None,
            }
        })
}

fn complete_procedure(
    pd: &ProcedureDeclaration,
    position: usize,
    table: &GlobalTable,
) -> Option<Vec<CompletionItem>> {
    pd.info
        .tokens
        .token_before(position)
        .and_then(|last_token| {
            let in_signature = pd
                .info
                .tokens
                .iter()
                .find(|token| matches!(token.token_type, TokenType::RParen | TokenType::LCurly))
                .map_or(true, |procedure_end_token| {
                    position < procedure_end_token.range.start
                });

            if in_signature {
                match last_token.token_type {
                    TokenType::LParen | TokenType::Comma => Some(vec![items::r#ref()]),
                    TokenType::Colon => Some(search_types(table)),
                    _ => None,
                }
            } else {
                // handle procedure body
                let lookup_table = LookupTable {
                    local_table: super::get_local_table(pd, table),
                    global_table: Some(table),
                };
                let in_statements = pd
                    .statements
                    .iter()
                    .find(|stmt| {
                        // prevent that error or empty statements disturb the calculation of
                        // the first real statement
                        !matches!(stmt, Statement::Error(_) | Statement::Empty(_))
                    })
                    .map_or(false, |first_stmt| position >= first_stmt.to_range().start);

                if in_statements {
                    complete_statements(&pd.statements, position, last_token, &lookup_table)
                } else {
                    // in variable declarations
                    match last_token.token_type {
                        TokenType::Colon => Some(search_types(table)),
                        TokenType::Semic | TokenType::LCurly => {
                            let completions =
                                vec![vec![snippets::var(), items::var()], new_stmt(&lookup_table)]
                                    .concat();
                            Some(completions)
                        }
                        _ => None,
                    }
                }
            }
        })
}

fn complete_statements(
    stmts: &[Statement],
    position: usize,
    last_token: &Token,
    lookup_table: &LookupTable,
) -> Option<Vec<CompletionItem>> {
    let mut last_stmt_is_if = false;
    stmts
        .iter()
        .find(|stmt| {
            if stmt.to_range().contains(&position) {
                true
            } else {
                last_stmt_is_if = matches!(stmt, Statement::If(_));
                false
            }
        })
        .map_or_else(
            || Some(new_stmt(lookup_table)),
            |stmt| complete_statement(stmt, position, last_token, last_stmt_is_if, lookup_table),
        )
}

fn complete_statement(
    stmt: &Statement,
    position: usize,
    last_token: &Token,
    last_stmt_is_if: bool,
    lookup_table: &LookupTable,
) -> Option<Vec<CompletionItem>> {
    // provide completion support for `else` after if statement
    if last_stmt_is_if && matches!(last_token.token_type, TokenType::RCurly) {
        let completions = vec![
            vec![snippets::r#else(), items::r#else()],
            new_stmt(lookup_table),
        ]
        .concat();
        return Some(completions);
    }

    match stmt {
        Statement::Assignment(a) => {
            let assign_token = a
                .info
                .tokens
                .iter()
                .find(|token| matches!(token.token_type, TokenType::Assign))
                .expect("Cannot be recognized as an assignment without assign token.");
            // before assign is only an arbitrary identifier
            if position >= assign_token.range.end {
                if let Some(local_table) = lookup_table.local_table {
                    return Some(search_variables(local_table));
                }
            }
        }
        Statement::Block(b) => {
            return complete_statements(&b.statements, position, last_token, lookup_table);
        }
        Statement::Call(c) => {
            let lparen_token = c
                .info
                .tokens
                .iter()
                .find(|token| matches!(token.token_type, TokenType::LParen))
                .expect("Cannot be recognized as a call statement without lparen token.");
            // procedure name completion is already provided by the
            // context based completions
            if position >= lparen_token.range.end {
                if let Some(local_table) = lookup_table.local_table {
                    return Some(search_variables(local_table));
                }
            }
        }
        Statement::If(i) => {
            if let Some(if_branch) = &i.if_branch {
                if if_branch.to_range().contains(&position) {
                    return complete_statement(
                        if_branch,
                        position,
                        last_token,
                        false,
                        lookup_table,
                    );
                }
            }
            if let Some(else_branch) = &i.else_branch {
                if else_branch.to_range().contains(&position) {
                    return complete_statement(
                        else_branch,
                        position,
                        last_token,
                        false,
                        lookup_table,
                    );
                }
            }
            if let Some(lparen_token) = &i
                .info
                .tokens
                .iter()
                .find(|token| matches!(token.token_type, TokenType::LParen))
            {
                if position >= lparen_token.range.end {
                    if let Some(local_table) = lookup_table.local_table {
                        return Some(search_variables(local_table));
                    }
                }
            }
        }
        Statement::While(w) => {
            if let Some(if_branch) = &w.statement {
                if if_branch.to_range().contains(&position) {
                    return complete_statement(
                        if_branch,
                        position,
                        last_token,
                        false,
                        lookup_table,
                    );
                }
            }
            if let Some(lparen_token) = &w
                .info
                .tokens
                .iter()
                .find(|token| matches!(token.token_type, TokenType::LParen))
            {
                if position >= lparen_token.range.end {
                    if let Some(local_table) = lookup_table.local_table {
                        return Some(search_variables(local_table));
                    }
                }
            }
        }
        Statement::Error(_) | Statement::Empty(_) => return Some(new_stmt(lookup_table)),
    }
    None
}

/// If the cursor is after a token, it should count as if it was on it.
const fn correct_index(index: usize) -> usize {
    if index > 0 {
        index - 1
    } else {
        0
    }
}

fn search_types(table: &GlobalTable) -> Vec<CompletionItem> {
    table
        .entries
        .iter()
        .filter(|(_, entry)| matches!(entry, GlobalEntry::Type(_)))
        .map(|(ident, entry)| CompletionItem {
            label: ident.clone(),
            kind: Some(CompletionItemKind::STRUCT),
            detail: Some(entry.to_string()),
            documentation: entry.doc().map(|docu| {
                Documentation::MarkupContent(lsp_types::MarkupContent {
                    kind: MarkupKind::Markdown,
                    value: docu.trim_start().to_string(),
                })
            }),
            ..Default::default()
        })
        .collect()
}

fn search_variables(table: &LocalTable) -> Vec<CompletionItem> {
    table
        .entries
        .iter()
        .filter(|(_, entry)| matches!(entry, LocalEntry::Variable(_)))
        .map(|(ident, entry)| CompletionItem {
            label: ident.clone(),
            kind: Some(CompletionItemKind::VARIABLE),
            detail: Some(entry.to_string()),
            documentation: entry.doc().map(|docu| {
                Documentation::MarkupContent(lsp_types::MarkupContent {
                    kind: MarkupKind::Markdown,
                    value: docu.trim_start().to_string(),
                })
            }),
            ..Default::default()
        })
        .collect()
}

fn search_procedures(table: &GlobalTable) -> Vec<CompletionItem> {
    table
        .entries
        .iter()
        .filter(|(_, entry)| matches!(entry, GlobalEntry::Procedure(_)))
        .map(|(ident, entry)| CompletionItem {
            label: ident.clone(),
            kind: Some(CompletionItemKind::FUNCTION),
            detail: Some(entry.to_string()),
            documentation: entry.doc().map(|docu| {
                Documentation::MarkupContent(lsp_types::MarkupContent {
                    kind: MarkupKind::Markdown,
                    value: docu.trim_start().to_string(),
                })
            }),
            ..Default::default()
        })
        .collect()
}

fn new_stmt(lookup_table: &LookupTable) -> Vec<CompletionItem> {
    vec![
        vec![
            snippets::r#if(),
            snippets::r#while(),
            items::r#if(),
            items::r#while(),
        ],
        lookup_table
            .local_table
            .map_or_else(Vec::new, search_variables),
        lookup_table
            .global_table
            .map_or_else(Vec::new, search_procedures),
    ]
    .concat()
}

mod items {
    use lsp_types::{CompletionItem, CompletionItemKind};
    use spl_frontend::lexer::token;

    macro_rules! item {
        ($name:ident, $label:expr) => {
            pub(super) fn $name() -> CompletionItem {
                CompletionItem {
                    label: $label.to_string(),
                    kind: Some(CompletionItemKind::KEYWORD),
                    ..Default::default()
                }
            }
        };
    }

    item!(array, token::ARRAY);
    item!(of, token::OF);
    item!(r#if, token::IF);
    item!(r#else, token::ELSE);
    item!(r#while, token::WHILE);
    item!(r#type, token::TYPE);
    item!(proc, token::PROC);
    item!(var, token::VAR);
    item!(r#ref, token::REF);

    /// int is a type, not a keyword
    pub(super) fn int() -> CompletionItem {
        CompletionItem {
            label: "int".to_string(),
            kind: Some(CompletionItemKind::STRUCT),
            ..Default::default()
        }
    }
}

mod snippets {
    use lsp_types::{CompletionItem, CompletionItemKind, InsertTextFormat};

    macro_rules! snippet {
        ($name:ident, $label:expr, $text:expr) => {
            pub(super) fn $name() -> CompletionItem {
                CompletionItem {
                    label: $label.to_string(),
                    kind: Some(CompletionItemKind::SNIPPET),
                    insert_text: Some($text.to_string()),
                    insert_text_format: Some(InsertTextFormat::SNIPPET),
                    ..Default::default()
                }
            }
        };
        ($name:ident, $text:expr) => {
            snippet!($name, stringify!($name), $text);
        };
    }

    snippet!(main, "proc main() {\n    $0\n}");
    snippet!(array, "array [$1] of $0");
    snippet!(proc, "proc $1($2) {\n    $0\n}");
    snippet!(var, "var $1: $0;");
    snippet!(r#type, "type", "type $1 = $0;");
    snippet!(r#if, "while", "while ($1) {\n    $0\n}");
    snippet!(r#while, "if", "if ($1) {\n    $0\n}");
    snippet!(r#else, "else", "else {\n    $0\n}");
}
