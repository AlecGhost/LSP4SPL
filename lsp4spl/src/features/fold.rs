use crate::document::{convert_range, DocumentRequest};
use color_eyre::eyre::Result;
use lsp_types::{FoldingRange, FoldingRangeKind, FoldingRangeParams};
use spl_frontend::{ast::GlobalDeclaration, ToRange};
use tokio::sync::mpsc::Sender;

pub(crate) async fn fold(
    doctx: Sender<DocumentRequest>,
    params: FoldingRangeParams,
) -> Result<Vec<FoldingRange>> {
    let doc_params = params.text_document;
    if let Some(doc_info) = super::get_doc_info(doc_params.uri, doctx).await? {
        let folding_ranges = doc_info
            .ast
            .global_declarations
            .iter()
            .filter_map(|gd| match gd {
                GlobalDeclaration::Procedure(p) => Some(p),
                _ => None,
            })
            .map(|p| {
                let range = convert_range(&p.to_range(), &doc_info.text);
                FoldingRange {
                    start_line: range.start.line,
                    end_line: range.end.line,
                    kind: Some(FoldingRangeKind::Region),
                    ..Default::default()
                }
            })
            .collect();
        Ok(folding_ranges)
    } else {
        Ok(Vec::new())
    }
}
