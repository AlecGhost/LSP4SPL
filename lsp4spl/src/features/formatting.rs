use crate::document::{convert_range, DocumentRequest};
use color_eyre::eyre::Result;
use fmt::Format;
use lsp_types::{DocumentFormattingParams, TextEdit};
use spl_frontend::AnalyzedSource;
use tokio::sync::mpsc::Sender;

pub async fn format(
    doctx: Sender<DocumentRequest>,
    params: DocumentFormattingParams,
) -> Result<Option<Vec<TextEdit>>> {
    let uri = params.text_document.uri;
    let options = params.options;
    if let Some(AnalyzedSource { ast, text, .. }) = super::get_doc_info(uri, doctx).await? {
        let formatting_options = if options.insert_spaces {
            fmt::FormattingOptions::new(' ', options.tab_size as usize)
        } else {
            fmt::FormattingOptions::new('\t', 1)
        };
        let new_text = ast.fmt(&formatting_options);
        if new_text == text {
            Ok(None)
        } else {
            let range = convert_range(&(0..text.len()), &text);
            Ok(Some(vec![TextEdit { range, new_text }]))
        }
    } else {
        Ok(None)
    }
}

mod fmt {
    use spl_frontend::{
        ast::{
            BlockStatement, GlobalDeclaration, IfStatement, ProcedureDeclaration, Program,
            Statement, TypeDeclaration, WhileStatement,
        },
        token::{Token, TokenType},
    };

    #[derive(Clone, Debug)]
    pub struct FormattingOptions {
        indent_symbol: char,
        indent_depth: usize,
    }

    impl FormattingOptions {
        pub const fn new(indent_symbol: char, indent_depth: usize) -> Self {
            Self {
                indent_symbol,
                indent_depth,
            }
        }

        pub fn indentation(&self) -> String {
            let indentation_characters = vec![self.indent_symbol; self.indent_depth];
            String::from_iter(indentation_characters)
        }
    }

    fn indent(text: String, f: &FormattingOptions) -> String {
        text.lines()
            .map(|line| f.indentation() + line + "\n")
            .collect()
    }

    fn add_leading_comments(text: String, tokens: &[Token]) -> String {
        tokens
            .iter()
            .map_while(|token| match &token.token_type {
                TokenType::Comment(comment) => Some(String::new() + "// " + comment.trim() + "\n"),
                _ => None,
            })
            .collect::<String>()
            + &text
    }

    fn add_all_comments(text: String, tokens: &[Token]) -> String {
        tokens
            .iter()
            .filter_map(|token| match &token.token_type {
                TokenType::Comment(comment) => Some(String::new() + "// " + comment.trim() + "\n"),
                _ => None,
            })
            .collect::<String>()
            + &text
    }

    pub trait Format {
        fn fmt(&self, f: &FormattingOptions) -> String;
    }

    impl Format for Program {
        fn fmt(&self, f: &FormattingOptions) -> String {
            self.global_declarations
                .iter()
                .map(|gd| gd.fmt(f))
                .reduce(|acc, gd| acc + "\n" + &gd)
                .unwrap_or_default()
        }
    }

    impl Format for GlobalDeclaration {
        fn fmt(&self, f: &FormattingOptions) -> String {
            match self {
                Self::Type(td) => td.fmt(f),
                Self::Procedure(pd) => pd.fmt(f),
                Self::Error(info) => info
                    .tokens
                    .iter()
                    .map(|token| token.to_string())
                    .reduce(|acc, token| {
                        if acc.ends_with('\n') {
                            acc + &token
                        } else {
                            acc + " " + &token
                        }
                    })
                    .unwrap_or_default(),
            }
        }
    }

    impl Format for ProcedureDeclaration {
        fn fmt(&self, f: &FormattingOptions) -> String {
            let name = fmt_or_empty(&self.name);
            let param_vec: Vec<String> = self
                .parameters
                .iter()
                .map(|param| add_all_comments(param.to_string(), &param.info.tokens))
                .collect();
            let params = if param_vec.is_empty() {
                String::new()
            } else {
                let len = param_vec.len();
                let has_comments = param_vec.iter().any(|param| param.contains("//"));
                // if there are more than three parameters,
                // or any of them is commented,
                // parameters should be in separate lines and indented
                if len > 3 || has_comments {
                    let params = param_vec
                        .into_iter()
                        .reduce(|acc, param| acc + ",\n" + param.as_str())
                        .expect("Params not empty was checked before");
                    "\n".to_string() + &indent(params, f)
                } else {
                    param_vec
                        .into_iter()
                        .reduce(|acc, param| acc + ", " + param.as_str())
                        .expect("Params not empty was checked before")
                }
            };
            let var_decs: String = self
                .variable_declarations
                .iter()
                .map(|var_dec| add_all_comments(var_dec.to_string(), &var_dec.info.tokens))
                .collect();
            let var_decs = indent(var_decs, f);
            let stmts: String = self.statements.iter().map(|stmt| stmt.fmt(f)).collect();
            let stmts = indent(stmts, f);
            let pd = match (var_decs.is_empty(), stmts.is_empty()) {
                (true, true) => format!("proc {}({}) {{}}\n", name, params),
                (true, false) => format!("proc {}({}) {{\n{}}}\n", name, params, stmts),
                (false, true) => format!("proc {}({}) {{\n{}}}\n", name, params, var_decs),
                (false, false) => {
                    format!("proc {}({}) {{\n{}\n{}}}\n", name, params, var_decs, stmts)
                }
            };
            add_leading_comments(pd, &self.info.tokens)
        }
    }

    impl Format for TypeDeclaration {
        fn fmt(&self, _f: &FormattingOptions) -> String {
            let type_expr = fmt_or_empty(&self.type_expr);
            let td = self.name.as_ref().map_or_else(
                || format!("type = {};\n", type_expr),
                |name| format!("type {} = {};\n", name, type_expr),
            );
            add_leading_comments(td, &self.info.tokens)
        }
    }

    impl Format for Statement {
        fn fmt(&self, f: &FormattingOptions) -> String {
            match self {
                Self::Assignment(a) => add_all_comments(a.to_string(), &a.info.tokens),
                Self::Block(b) => b.fmt(f),
                Self::Call(c) => add_all_comments(c.to_string(), &c.info.tokens),
                Self::If(i) => i.fmt(f),
                Self::While(w) => w.fmt(f),
                Self::Empty(info) => add_all_comments(";\n".to_string(), &info.tokens),
                Self::Error(_) => String::new(),
            }
        }
    }

    impl Format for BlockStatement {
        fn fmt(&self, f: &FormattingOptions) -> String {
            let stmt = if self.statements.is_empty() {
                "{}\n".to_string()
            } else {
                let stmts: String = self.statements.iter().map(|stmt| stmt.fmt(f)).collect();
                let stmts = indent(stmts, f);
                format!("{{\n{}}}\n", stmts)
            };
            add_leading_comments(stmt, &self.info.tokens)
        }
    }

    impl Format for IfStatement {
        fn fmt(&self, f: &FormattingOptions) -> String {
            let condition = fmt_or_empty(&self.condition);
            let stmt = if self.else_branch.is_some() {
                format!(
                    "if ({}){}else{}",
                    condition,
                    fmt_branch(f, &self.if_branch, ' '),
                    fmt_branch(f, &self.else_branch, '\n'),
                )
            } else {
                format!("if ({}){}", condition, fmt_branch(f, &self.if_branch, '\n'))
            };
            add_leading_comments(stmt, &self.info.tokens)
        }
    }

    impl Format for WhileStatement {
        fn fmt(&self, f: &FormattingOptions) -> String {
            let condition = fmt_or_empty(&self.condition);
            let stmt = format!(
                "while ({}){}",
                condition,
                fmt_branch(f, &self.statement, '\n')
            );
            add_leading_comments(stmt, &self.info.tokens)
        }
    }

    fn fmt_branch(f: &FormattingOptions, branch: &Option<Box<Statement>>, ending: char) -> String {
        branch.as_ref().map_or_else(
            || format!("{}", ending),
            |stmt| {
                if let Statement::Block(b) = stmt.as_ref() {
                    // similar to `BlockStatement::fmt`,
                    // except that a space is inserted before the brackets
                    // and if there are statements inside the block,
                    // `ending` is appended instead of a newline
                    if b.statements.is_empty() {
                        " {}\n".to_string()
                    } else {
                        let stmts: String = b.statements.iter().map(|stmt| stmt.fmt(f)).collect();
                        let stmts = indent(stmts, f);
                        format!(" {{\n{}}}{}", stmts, ending)
                    }
                } else {
                    let stmt = indent(stmt.fmt(f), f);
                    format!("\n{}", stmt)
                }
            },
        )
    }

    fn fmt_or_empty<T: std::fmt::Display>(opt: &Option<T>) -> String {
        opt.as_ref()
            .map_or_else(String::new, |inner| inner.to_string())
    }
}
