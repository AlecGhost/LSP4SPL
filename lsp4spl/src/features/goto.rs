use super::DocumentCursor;
use crate::document::{convert_range, DocumentRequest};
use color_eyre::eyre::Result;
use lsp_types::{
    request::{GotoDeclarationParams, GotoImplementationParams, GotoTypeDefinitionParams},
    GotoDefinitionParams, Location,
};
use spl_frontend::{
    table::{DataType, Entry, LookupTable, Table},
    ToRange,
};
use tokio::sync::mpsc::Sender;

pub(crate) async fn declaration(
    doctx: Sender<DocumentRequest>,
    params: GotoDeclarationParams,
) -> Result<Option<Location>> {
    let doc_params = params.text_document_position_params;
    let uri = doc_params.text_document.uri.clone();
    if let Some(cursor) = super::doc_cursor(doc_params, doctx).await? {
        if let Some(ident) = &cursor.ident() {
            let DocumentCursor {
                doc_info, context, ..
            } = cursor;
            if let Some(entry) = context {
                match &entry {
                    Entry::Type(_) => {
                        if let Some(entry) = doc_info.table.lookup(&ident.value) {
                            return Ok(Some(Location {
                                uri,
                                range: convert_range(
                                    &entry.to_range(),
                                    &doc_info.text,
                                ),
                            }));
                        }
                    }
                    Entry::Procedure(p) => {
                        let lookup_table = LookupTable {
                            global_table: &doc_info.table,
                            local_table: &p.local_table,
                        };
                        if let Some(entry) = lookup_table.lookup(&ident.value) {
                            return Ok(Some(Location {
                                uri,
                                range: convert_range(
                                    &entry.to_range(),
                                    &doc_info.text,
                                ),
                            }));
                        }
                    }
                    Entry::Variable(v) => {
                        log::error!("Found illegal variable in global table {:#?}", v);
                        panic!("Found illegal variable in global table {:#?}", v);
                    }
                }
            }
        }
    }
    Ok(None)
}

/// Calls `goto::declaration` because in SPL, there is no conceptual difference
/// between declaration and definition
pub(crate) async fn definition(
    doctx: Sender<DocumentRequest>,
    params: GotoDefinitionParams,
) -> Result<Option<Location>> {
    declaration(doctx, params).await
}

pub(crate) async fn type_definition(
    doctx: Sender<DocumentRequest>,
    params: GotoTypeDefinitionParams,
) -> Result<Option<Location>> {
    let doc_params = params.text_document_position_params;
    let uri = doc_params.text_document.uri.clone();
    if let Some(cursor) = super::doc_cursor(doc_params, doctx).await? {
        if let Some(ident) = &cursor.ident() {
            let DocumentCursor {
                doc_info, context, ..
            } = cursor;
            if let Some(entry) = context {
                match &entry {
                    Entry::Type(_) => {
                        if let Some(entry) = doc_info.table.lookup(&ident.value) {
                            match &entry {
                                Entry::Type(_) => {
                                    return Ok(Some(Location {
                                        uri,
                                        range: convert_range(
                                            &entry.to_range(),
                                            &doc_info.text,
                                        ),
                                    }));
                                }
                                Entry::Procedure(_) => { /* no type definition */ }
                                Entry::Variable(v) => {
                                    log::error!("Found illegal variable in global table {:#?}", v);
                                    panic!("Found illegal variable in global table {:#?}", v);
                                }
                            }
                        }
                    }
                    Entry::Procedure(p) => {
                        let lookup_table = LookupTable {
                            global_table: &doc_info.table,
                            local_table: &p.local_table,
                        };
                        if let Some(entry) = lookup_table.lookup(&ident.value) {
                            match &entry {
                                Entry::Type(_) => {
                                    return Ok(Some(Location {
                                        uri,
                                        range: convert_range(
                                            &entry.to_range(),
                                            &doc_info.text,
                                        ),
                                    }));
                                }
                                Entry::Procedure(_) => { /* no type definition */ }
                                Entry::Variable(v) => {
                                    if let Some(DataType::Array { creator, .. }) = &v.data_type {
                                        return Ok(Some(Location {
                                            uri,
                                            range: convert_range(
                                                &creator.to_range(),
                                                &doc_info.text,
                                            ),
                                        }));
                                    }
                                    /* cannot look up primitive types */
                                }
                            }
                        }
                    }
                    Entry::Variable(v) => {
                        log::error!("Found illegal variable in global table {:#?}", v);
                        panic!("Found illegal variable in global table {:#?}", v);
                    }
                }
            }
        }
    }
    Ok(None)
}

/// Essentially the same as `goto::declaration`, but only for procedures
pub(crate) async fn implementation(
    doctx: Sender<DocumentRequest>,
    params: GotoImplementationParams,
) -> Result<Option<Location>> {
    let doc_params = params.text_document_position_params;
    let uri = doc_params.text_document.uri.clone();
    if let Some(cursor) = super::doc_cursor(doc_params, doctx).await? {
        if let Some(ident) = &cursor.ident() {
            let DocumentCursor {
                doc_info, context, ..
            } = cursor;
            if let Some(entry) = context {
                match &entry {
                    Entry::Procedure(p) => {
                        let lookup_table = LookupTable {
                            global_table: &doc_info.table,
                            local_table: &p.local_table,
                        };
                        if let Some(entry) = lookup_table.lookup(&ident.value) {
                            if let Entry::Procedure(_) = entry {
                                return Ok(Some(Location {
                                    uri,
                                    range: convert_range(
                                        &entry.to_range(),
                                        &doc_info.text,
                                    ),
                                }));
                            }
                            /* no implementation for types and variables */
                        }
                    }
                    Entry::Type(_) => { /* no implementation for types */ }
                    Entry::Variable(v) => {
                        log::error!("Found illegal variable in global table {:#?}", v);
                        panic!("Found illegal variable in global table {:#?}", v);
                    }
                }
            }
        }
    }
    Ok(None)
}
