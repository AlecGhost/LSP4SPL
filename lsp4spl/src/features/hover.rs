use super::DocumentPrelude;
use crate::document::{convert_range, DocumentRequest};
use color_eyre::eyre::Result;
use lsp_types::{Hover, HoverContents, HoverParams, MarkupContent, MarkupKind, Range};
use spl_frontend::table::{Entry, LookupTable, Table};
use tokio::sync::mpsc::Sender;

fn create_hover(entry: &Entry, range: Range) -> Hover {
    Hover {
        contents: HoverContents::Markup(MarkupContent {
            kind: MarkupKind::Markdown,
            value: entry.to_string(),
        }),
        range: Some(range),
    }
}

pub(crate) async fn hover(
    doctx: Sender<DocumentRequest>,
    params: HoverParams,
) -> Result<Option<Hover>> {
    let doc_params = params.text_document_position_params;
    if let Some(DocumentPrelude {
        doc_info,
        ident,
        entry,
    }) = super::document_prelude(doc_params, doctx).await?
    {
        match &entry {
            Entry::Type(_) => {
                if let Some(ranged_entry) = doc_info.table.lookup(&ident) {
                    return Ok(Some(create_hover(
                        &ranged_entry.entry,
                        convert_range(&ident.range, &doc_info.text),
                    )));
                }
            }
            Entry::Procedure(p) => {
                let lookup_table = LookupTable {
                    global_table: &doc_info.table,
                    local_table: &p.local_table,
                };
                if let Some(ranged_entry) = lookup_table.lookup(&ident) {
                    return Ok(Some(create_hover(
                        &ranged_entry.entry,
                        convert_range(&ident.range, &doc_info.text),
                    )));
                }
            }
            Entry::Variable(v) => {
                log::error!("Found illegal variable in global table {:#?}", v);
                panic!("Found illegal variable in global table {:#?}", v);
            }
        }
    }
    Ok(None)
}
