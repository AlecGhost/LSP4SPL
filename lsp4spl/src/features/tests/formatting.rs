use super::*;
use crate::features::format;
use insta;
use lsp_types::{
    DocumentFormattingParams, FormattingOptions, TextDocumentIdentifier, WorkDoneProgressParams,
};

async fn test_formatting(text: &str) -> String {
    let uri = Url::parse("file:///test.spl").unwrap();
    let params = DocumentFormattingParams {
        text_document: TextDocumentIdentifier { uri: uri.clone() },
        options: FormattingOptions::default(),
        work_done_progress_params: WorkDoneProgressParams::default(),
    };
    let result: String = test_feature(format, uri.clone(), text, params)
        .await
        .unwrap()
        .unwrap_or_default()
        .first()
        .map(|text_edit| text_edit.new_text.clone())
        .unwrap_or_default();
    result
}

#[tokio::test]
async fn empty() {
    let result = test_formatting("").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn type_dec_correct() {
    let result = test_formatting("type a = int;\n").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn type_dec_incorrect() {
    let result = test_formatting("type\n a=int;").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn type_dec_missing_semic() {
    let result = test_formatting("type a=int").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn proc_dec_correct() {
    let result = test_formatting("proc a() {}\n").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn proc_dec_no_body() {
    let result = test_formatting("proc a() {\n}").await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn acker() {
    let acker = std::fs::read_to_string("tests/features/formatting/acker.spl").unwrap();
    let result = test_formatting(&acker).await;
    insta::assert_snapshot!(result);
}

#[tokio::test]
async fn big_test() {
    let big_test = std::fs::read_to_string("tests/features/formatting/bigtest.spl").unwrap();
    let result = test_formatting(&big_test).await;
    insta::assert_snapshot!(result);
}
