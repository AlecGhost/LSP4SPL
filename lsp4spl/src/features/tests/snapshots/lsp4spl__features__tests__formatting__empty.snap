---
source: lsp4spl/src/features/tests/formatting.rs
expression: result
---

